tests/mock_crust/mod.rs

```diff
@@ -245,3 +245,111 @@ fn multiple_clients_per_proxy() {
+    let leaky_rate = 2 * CLIENT_GET_CHARGE;
+    let per_client_cap = CAPACITY / MAX_CLIENTS_PER_PROXY as u64;
+    for i in 0..10 {
+        trace!("iteration {:?}", i);
+        let mut clients_sent = HashMap::new();
+        for (j, client) in clients.iter_mut().enumerate() {
+            if rng.gen_weighted_bool(2) {
+                let msg_id = MessageId::new();
+                unwrap!(client.inner.get_idata(dst, data_id, msg_id));
+                let _ =
+                    clients_sent.insert(msg_id,
+                                        IpAddr::from([0, 0, 0, (j + min_section_size) as u8]));
+            }
+        }
+        trace!("clients_sent: {:?}", clients_sent);
+        let _ = poll_all(&mut nodes, &mut clients);
+
+        let mut request_received: HashMap<MessageId, usize> = HashMap::new();
+        for node in nodes.iter_mut().filter(|n| n.is_recipient(&dst)) {
+            while let Ok(event) = node.try_next_ev() {
+                if let Event::Request {
+                           request: Request::GetIData { msg_id: req_message_id, .. }, ..
+                       } = event {
+                    let entry = request_received
+                        .entry(req_message_id)
+                        .or_insert_with(|| 0);
+                    *entry += 1;
+                }
+            }
+        }
+        trace!("request_received: {:?}", request_received);
+
+        for (msg_id, count) in &request_received {
+            assert_eq!(*count, min_section_size);
+            let _ = unwrap!(clients_sent.remove(msg_id));
+            total_usage += CLIENT_GET_CHARGE;
+        }
+        assert!(total_usage <= CAPACITY);
+
+
+        let clients_usage = nodes[0].inner.get_clients_usage();
+        assert_eq!(clients_usage.iter()
+                             .filter(|&(_, usage)| *usage > per_client_cap)
+                             .count(),
+                   0);
+        for ip in clients_sent.values() {
+            assert!(unwrap!(clients_usage.get(ip)) + CLIENT_GET_CHARGE > per_client_cap);
+        }
+
+        FakeClock::advance_time(wait_millis);
+        total_usage -= leaky_rate;
+    }
+}
+
+#[test]
+/// Connect a client to the network then send an invalid message.
+/// Expect the client will be banned (disconnected);
+fn ban_malicious_client() {
+    let min_section_size = 8;
+    let network = Network::new(min_section_size, None);
+    let mut nodes = create_connected_nodes(&network, min_section_size);
+    let mut clients = create_connected_clients(&network, &mut nodes, 1);
+
+
+
+
+    let mut rng = network.new_rng();
+    let data_name: XorName = rng.gen();
+    let group_auth = Authority::NaeManager(data_name);
+    let client_auth = Authority::Client {
+        client_id: *clients[0].full_id.public_id(),
+        proxy_node_name: nodes[0].name(),
+    };
+    let msg_id = MessageId::new();
+
+    // Send a get request from NM to client, which makes client sends an ACK to proxy.
```

Comment:
This seems like quite a roundabout way to make the client send an invalid message!  I guess I'd make it more explicit by having the client send one of each of the variants of `HopMessage` and `DirectMessage` it's not allowed, and also one of each of the `HopMessage`s it _is_ allowed but with an invalid priority applied.  Also it'd be good to have a similar test where the client bootstraps and doesn't send the expected `BootstrapRequest` (either sends an invalidly-signed one or sends a different variant).  There are more malicious scenarios we've catered for - I guess they should all be tested really.